Adding optional run/hit/error display, and a config option for this.

mmendelson222 opened the pull request:
Created a configuration option to turn on/off display of runs and hits, in addition to score.  To activate, just add this to your configuration:  "game": {"show_hits_errors": true},

The rendering was a little tricky - the three scores right-align with the [teams][runs] coordinate the same way it works with runs.  I added logic to reduce the space between the three numbers since spaces look too wide.  Width varies depending on the number of digits in each number, but the two teams scores stay aligned.  For example:

![runs_hits_errors](https://user-images.githubusercontent.com/1591186/135737262-39ec2408-e1a2-4f31-9dc2-c4383f9c80c9.JPG)
 
I ran isort and black on changed files :)

WardBrian commented on renderers/games/teams.py:
```diff
@@ -78,11 +85,38 @@ def __render_team_text(canvas, layout, colors, team, homeaway, full_team_names,
-    graphics.DrawText(canvas, font["font"], team_runs_x, coords["y"], text_color_graphic, team_runs)
+    font = layout.font(f"teams.runs.{homeaway}")  # just use the runs font
+    component_val = str(component_val)
+    std_spacing = True
```
Unused flag?

mmendelson222 replied:
Yup, will remove.

WardBrian commented on renderers/games/teams.py:
```diff
@@ -78,11 +85,38 @@ def __render_team_text(canvas, layout, colors, team, homeaway, full_team_names,
+    font = layout.font(f"teams.runs.{homeaway}")  # just use the runs font
+    component_val = str(component_val)
+    std_spacing = True
+    for i in range(width):
+        score_x = coords["x"] - font["size"]["width"] + 1
+        coords["x"] = score_x
+        if len(component_val) > i:
+            c = component_val[len(component_val) - i - 1]
+            graphics.DrawText(canvas, font["font"], score_x, coords["y"], text_color_graphic, c)
```
Can you explain this loop? Why is the `DrawText` call inside an if?

WardBrian replied:
Also, is the return value of this function different than the score_spacing value? Can we not just re-use that?

mmendelson222 replied:
In retrospect my logic here might be more complex than it needs to be.  Also I didn't really test with scores in the 20's, and I see that the spacing is different between 1's and 2's.  I would like to work out different spacing for 10's and 20's.  I think I will take some more time and refactor this a bit.

WardBrian replied:
Is the font really a different width, or does it just have different values?

The nice thing is for testing you can just have game.away_hits() return 22 or something

mmendelson222 replied:
The font is fixed width.  All characters are 3 pixels wide with the exception of the "1" which is drawn in the leftmost two pixels.  So if you're rendering a 10, it looks like there's more space between the 1 and the 0.  I'm fixing this by shifting the 1 over (but only if it's in the 10's column).  Whew!

WardBrian replied:
I think this change isn't a good idea. It complicates the code and looks odd if the font used isnt 4x6 (for example, try the 64x64 layout or 128x64). It's not idea for the 4x6 case but I think it's better to not special-case the teens.

WardBrian commented on renderers/games/teams.py:
```diff
@@ -78,11 +85,38 @@ def __render_team_text(canvas, layout, colors, team, homeaway, full_team_names,
+
+
+def __render_team_score(canvas, layout, colors, team, homeaway, default_colors, score_spacing, show_hits_errors):
+    if show_hits_errors:
+        coords = layout.coords(f"teams.runs.{homeaway}").copy()
```
```suggestion
    coords = layout.coords(f"teams.runs.{homeaway}").copy()
    if show_hits_errors:
```

If we copy in both branches we can just copy before. I'm not sure if I prefer this to just making local variables `x,y` and updating them/ passing to the render_component function, but this is probably fine.

mmendelson222 replied:
Yes that's right, I'll make this change.

WardBrian commented on renderers/games/teams.py:
```diff
@@ -78,11 +85,38 @@ def __render_team_text(canvas, layout, colors, team, homeaway, full_team_names,
+            canvas, layout, colors, homeaway, default_colors, coords, team.hits, score_spacing["hits"]
+        )
+        coords["x"] = x_calc - 2
+        x_calc = __render_score_component(
+            canvas, layout, colors, homeaway, default_colors, coords, team.runs, score_spacing["runs"]
+        )
+    else:
+        coords = layout.coords(f"teams.runs.{homeaway}").copy()
+        __render_score_component(
+            canvas, layout, colors, homeaway, default_colors, coords, team.runs, score_spacing["runs"]
+        )
```
```suggestion
    __render_score_component(
        canvas, layout, colors, homeaway, default_colors, coords, team.runs, score_spacing["runs"]
    )
```

Drawing runs shouldn't need to be conditional, right? You can just do it following the if

mmendelson222 replied:
Yup, right again!

WardBrian commented on config.json.example:
```diff
@@ -3,6 +3,9 @@
 		"teams": ["Cubs"],
 		"divisions": ["NL Central"]
 	},
+	"game": {
+		"show_hits_errors": false
+	},
```
Part of me is wondering if `rotation` should be a sub-grouping under game (or 'games')? It would break existing config and require a lot of changes to the config class' loading routine. What do you think?

mmendelson222 replied:
The configuration logic is very simple - everything just gets dumped into variables in config/__init__.py.  So moving things around would not be complicated.  It might break people's custom configurations though, if anyone is using this code.

WardBrian replied:
Yeah certainly not complicated, just a bunch of tiny changes. It also wouldn't be crazy to update an existing config, but for now I don't think it's worth it. Something to think about in the future

WardBrian commented on data/config/__init__.py:
```diff
@@ -25,6 +25,9 @@ def __init__(self, filename_base, width, height):
         self.preferred_divisions = json["preferred"]["divisions"]
 
+        # Game Screen Display
+        self.show_hits_errors = json["game"]["show_hits_errors"]
```
How does this look on 32x32 boards? We could manually override this to be false on those dimensions if needed.

The way I'm handling batters due up and the at bat/pitcher names ATM is just drawing them outside the dimensions of the board, but that's obviously not ideal and I should reconsider at some point.

WardBrian replied:
Okay it looks fine unless it any of the numbers hit double digits. We can probably let people disable it in their own config then.

mmendelson222 replied:
I did test this, and came to the conclusion that there's just not enough space when the numbers get high.  But it does work, and because this is optional there's no harm done.

WardBrian reviewed (changes requested):
This is really great. I've seen other people do this before but they never got the formatting right so that the columns all lined up if any of the numbers exceeded 9

pjockey commented:
Are these changes ready to go? I get a fatal error: `unable to access 'https://github.com/wardbrian/mlb-led-scoreboard/': Could not resolve host: github.com`

WardBrian commented:
@pjockey that seems like an unrelated problem, what prompted it?

pjockey commented:
> @pjockey that seems like an unrelated problem, what prompted it?

I was trying to use `git pull`

WardBrian commented:
git pull won't get you these changes yet, but it shouldn't error. Have you tried again?

pjockey commented:
> git pull won't get you these changes yet, but it shouldn't error. Have you tried again?

Still the same. I did an update and upgrade to the Pi and then rebooted it.

WardBrian commented:
@pjockey that honestly sounds like a firewall or other network issue. can you `ping github.com`?

pjockey commented:
> @pjockey that honestly sounds like a firewall or other network issue. can you `ping github.com`?

I had disable the IPV6 settings and that seemed to rendered the outside world non existent.  Switched back all good.

mmendelson222 commented:
OK, I have simplified the logic and added some comments so that's it's more understandable.  Also I've made some spacing accommodations - see my previous comment and my code comments.  I'm attaching a test config that has some nice high scores...
[config-high-score.json.txt](https://github.com/WardBrian/mlb-led-scoreboard/files/7274460/config-high-score.json.txt)

WardBrian commented:
Code looks a lot better, I would just personally ditch the extra logic around the 10s digit. I think the original style looks fine and it works a lot better for other fonts

WardBrian commented on renderers/games/teams.py:
```diff
@@ -78,11 +93,29 @@ def __render_team_text(canvas, layout, colors, team, homeaway, full_team_names,
-def __render_team_score(canvas, layout, colors, runs, homeaway, default_colors):
+def __render_score_component(canvas, layout, colors, homeaway, default_colors, coords, component_val, width):
+    # The coords passed in are the rightmost pixel.
+    RHE_SPACING = 3  # Number of pixels between runs/hits and hits/errors.
```
Can we make this either dependent on the font size or part of the config?

mmendelson222 replied:
Thanks - I appreciate how you push me toward simplicity.  I got rid of the extra logic around 1's and made spacing dependent on font size.  However I do like the reduced spacing between digits - it makes a visible difference especially because we're not using a full character's space between scores. 

What I've come up with is hopefully a compromise - a "compact" option - it looks good with 1 and ok with 2.  This way it's also possible to compare the two approaches.  Note that the configuration has changed. Let me know what you think!

Unfortunately black is now spreading some of the function calls across many lines.  Too many parameters, I guess.

WardBrian replied:
A nice compromise. I think this compact setting should live in the coordinate .jsons, not the global config. That way it can place nice with different sizes. We should update all the example sizes with a reasonable default - probably false on all the larger sizes, but it's worth seeing how it looks first

mmendelson222 replied:
That's certainly a possibility.  I didn't go this route because I think the coordinates file just contain x,y coodinates, and this is more of a spacing option.  Also my thought was that most people who implement this will be doing so for only one size, but I guess you never know. 

So in that case should we put BOTH settings in the layout? Perhaps something like this... 

runs_hits_errors {
show: true
compact: true
}

I like the idea of just using the "runs" coordinate as the location, but we could instead put "home" and "away" coordinates to the above.  What do you think?

WardBrian replied:
We already have flags in the coordinates in a few places (one of which was even there before my tinkering):

https://github.com/WardBrian/mlb-led-scoreboard/blob/9c8cbf0f1bb6f86e509ab47cf004b22345572f0a/coordinates/w64h32.json.example#L56-L57
https://github.com/WardBrian/mlb-led-scoreboard/blob/9c8cbf0f1bb6f86e509ab47cf004b22345572f0a/coordinates/w64h32.json.example#L198

So I think this is fine. Let's just put it in Runs, no reason to split home/away. I'd change 'compact' to something like 'compress_digits' or similar, but other than that it sounds good to me.

WardBrian commented on README.md:
````diff
@@ -173,6 +173,9 @@ A default `config.json.example` file is included for reference. Copy this file t
 ```
 
+### Additional Features
+* Runs/Hits/Errors - Hits are always shown on the games screen, but you can enable or adjust spacing of a "runs, hits, errors" display.  Take a look at the [coordinates readme file](/coordinates/README.md) for details. 
````
```suggestion
* Runs/Hits/Errors - Runs are always shown on the games screen, but you can enable or adjust spacing of a "runs, hits, errors" display.  Take a look at the [coordinates readme file](/coordinates/README.md) for details. 
```

mmendelson222 replied:
Readme correction

WardBrian commented on coordinates/w128h32.json.example:
```diff
@@ -269,6 +269,11 @@
       }
     },
     "runs": {
+      "runs_hits_errors": {
+        "show": false,
+        "compress_digits": false,
+        "spacing" : 3
+      },
```
Does 3 look okay if you turn on show, or is it just a dummy value?

mmendelson222 replied:
dummy value for now - that's why it's turned off. Do you know of anyone who is using this size?

WardBrian replied:
I don't know anyone personally, but there definitely are users. I'm happy with a dummy value I was just curious

WardBrian commented:
This is looking really good now, just a few last things. Thanks for putting up with so much nit-picking

mmendelson222 commented:
> This is looking really good now, just a few last things. Thanks for putting up with so much nit-picking

True, but I can relate.  Also I like how this is turning out.

mmendelson222 commented:
I'm hoping you're not going to say "test all resolutions in the emulator!"

WardBrian commented:
I just pushed a commit to fix a crash in this PR with the way layout states are handled. Made me realize a great application for this: turning on the extended display if/when there is a no hitter going on
